packages/core/src/deployment/Deployment.ts

```diff
@@ -1,177 +1,205 @@
+  public readExecutionErrors() {
+    return [...Object.entries(this.state.execution.vertexes)]
+      .filter(([_id, value]) => value.status === "FAILED")
+      .reduce(
+        (
+          acc: { [key: number]: VertexVisitResultFailure },
+          [id, { result }]
+        ) => {
+          if (
+            result === undefined ||
+            result === null ||
+            result._kind !== "failure"
+          ) {
+            return acc;
+          }
+
+          acc[parseInt(id, 10)] = result;
 
-    this._renderToUi(this.state);
+          return acc;
+        },
+        {}
+      );
   }
 
-  public updateCurrentBatchWithResult(
-    vertexId: number,
-    result: VertexVisitResult
-  ) {
-    log(`Update current with batch result for ${vertexId}`, result);
-    this.state = deployStateReducer(this.state, {
-      type: "UPDATE_CURRENT_BATCH_WITH_RESULT",
-      vertexId,
-      result,
-    });
+  public readExecutionHolds(): VertexDescriptor[] {
+    const executionGraph = this.state.transform.executionGraph;
 
-    this._renderToUi(this.state);
-  }
+    if (executionGraph === null) {
+      throw new Error("Cannot read from unset execution graph");
+    }
 
-  public readExecutionErrors() {
-    const errors = [...this.state.execution.errored].reduce(
-      (acc: { [key: number]: VertexVisitResultFailure }, id) => {
-        const result = this.state.execution.resultsAccumulator.get(id);
-
-        if (
-          result === undefined ||
-          result === null ||
-          result._kind === "success"
-        ) {
-          return acc;
+    return [...Object.entries(this.state.execution.vertexes)]
+      .filter(([_id, value]) => value.status === "HOLD")
+      .map(([id]) => {
+        const vertex = executionGraph.vertexes.get(parseInt(id, 10));
+
+        if (vertex === undefined) {
+          return null;
         }
 
-        acc[id] = result;
+        const descriptor: VertexDescriptor = {
+          id: vertex.id,
+          label: vertex.label,
+          type: vertex.type,
+        };
 
-        return acc;
-      },
-      {}
-    );
-
-    return errors;
+        return descriptor;
+      })
+      .filter((x): x is VertexDescriptor => Boolean(x));
   }
 
   public hasUnstarted(): boolean {
-    return this.state.execution.unstarted.size > 0;
+    return Object.values(this.state.execution.vertexes).some(
+      (v) => v.status === "UNSTARTED"
+    );
   }
 
   public hasErrors(): boolean {
-    return this.state.execution.errored.size > 0;
+    return Object.values(this.state.execution.vertexes).some(
+      (v) => v.status === "FAILED"
+    );
+  }
+
+  public hasHolds(): boolean {
+    return Object.values(this.state.execution.vertexes).some(
+      (v) => v.status === "HOLD"
+    );
+  }
+
+  private async _runDeploymentCommand(
+    logMessage: string | [string, any[]],
+    command: DeployStateCommand
+  ): Promise<void> {
+    log.apply(this, typeof logMessage === "string" ? [logMessage] : logMessage);
```

Comment:
I'm not sure I really love this as a solution, but I can't think of a better alternative right now either so it's fine

Reply:
I agree, we should settle for my mediocrity, while we come up with something better. 😄